File: windows/src/Notchle.Core/App/Spotify/SpotifyPlayerApi.cs
using System.Globalization;
using System.Net;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Notchle.Core.App.Playback;

namespace Notchle.Core.App.Spotify;

public sealed record SpotifyDevice(string? Id, string Name, string Type, bool IsActive, bool IsRestricted);

/// The parts of GET /v1/me/player the snippet timing needs.
/// `LinkedFromUri` is `item.linked_from.uri`: Spotify plays a relinked version of a track for the
/// account's market and reports the requested uri there, so it counts as the requested track.
public sealed record SpotifyPlayback(bool IsPlaying, int ProgressMs, string? ItemUri, string? CurrentlyPlayingType, string? DeviceId,
    string? LinkedFromUri = null)
{
    public PlaybackSample ToSample(string expectedUri) => new(
        Position: ProgressMs / 1000.0,
        IsPlaying: IsPlaying,
        WrongTrack: ItemUri != expectedUri && LinkedFromUri != expectedUri && CurrentlyPlayingType != "ad",
        Failure: CurrentlyPlayingType == "ad" && IsPlaying ? "Spotify is playing an ad" : null);
}

/// Spotify Web API player endpoints: pure request building, response parsing and error mapping.
/// https://developer.spotify.com/documentation/web-api/reference/start-a-users-playback
public static class SpotifyPlayerApi
{
    public static readonly Uri BaseUri = new("https://api.spotify.com/v1/");

    public static HttpRequestMessage Devices() => new(HttpMethod.Get, new Uri(BaseUri, "me/player/devices"));

    /// `market=from_token` so Spotify reports `linked_from` for relinked tracks.
    public static HttpRequestMessage PlaybackState() => new(HttpMethod.Get, new Uri(BaseUri, "me/player?market=from_token"));

    /// PUT /me/player: move playback to this PC's Spotify app without starting anything.
    public static HttpRequestMessage Transfer(string deviceId) =>
        Put("me/player", new JsonObject { ["device_ids"] = new JsonArray(deviceId), ["play"] = false });

    /// PUT /me/player/play: this one track, from positionMs, on the given device.
    /// With `contextUri` (the track's album) it plays "that album, from this track": measured on
    /// Tren's account (2026-09-24), a bare `uris` play is accepted with 204 but Spotify then
    /// empties the player, while the album-context play works. Keys are in sorted order, as on macOS.
    public static HttpRequestMessage Play(string deviceId, string trackUri, int positionMs, string? contextUri = null)
    {
        var path = $"me/player/play?device_id={Uri.EscapeDataString(deviceId)}";
        var position = Math.Max(0, positionMs);
        return contextUri is null
            ? Put(path, new JsonObject { ["position_ms"] = position, ["uris"] = new JsonArray(trackUri) })
            : Put(path, new JsonObject
            {
                ["context_uri"] = contextUri,
                ["offset"] = new JsonObject { ["uri"] = trackUri },
                ["position_ms"] = position,
            });
    }

    /// GET /tracks/{id}: used for the album uri (the play context).
    public static HttpRequestMessage Track(string id) =>
        new(HttpMethod.Get, new Uri(BaseUri, $"tracks/{Uri.EscapeDataString(id)}?market=from_token"));

    public static string? ParseAlbumUri(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.ValueKind == JsonValueKind.Object
                   && doc.RootElement.TryGetProperty("album", out var album) && album.ValueKind == JsonValueKind.Object
                ? Str(album, "uri")
                : null;
        }
        catch (JsonException) { return null; }
    }

    /// PUT /me/player/play without a body resumes the current track.
    public static HttpRequestMessage Resume(string? deviceId) => Put(WithDevice("me/player/play", deviceId), null);

    public static HttpRequestMessage Pause(string? deviceId) => Put(WithDevice("me/player/pause", deviceId), null);

    public static HttpRequestMessage Seek(int positionMs, string? deviceId) =>
        Put(WithDevice($"me/player/seek?position_ms={Math.Max(0, positionMs).ToString(CultureInfo.InvariantCulture)}", deviceId), null);

    public static IReadOnlyList<SpotifyDevice> ParseDevices(string json)
    {
        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("devices", out var devices) || devices.ValueKind != JsonValueKind.Array)
            return Array.Empty<SpotifyDevice>();
        return devices.EnumerateArray().Select(d => new SpotifyDevice(
            Id: Str(d, "id"),
            Name: Str(d, "name") ?? "",
            Type: Str(d, "type") ?? "",
            IsActive: Bool(d, "is_active"),
            IsRestricted: Bool(d, "is_restricted"))).ToList();
    }

    /// The Spotify desktop app on this PC: a controllable "Computer" device, preferring the one
    /// named after this machine (the app's default name), then the active one.
    public static SpotifyDevice? PickLocalDevice(IReadOnlyList<SpotifyDevice> devices, string machineName)
    {
        var candidates = devices
            .Where(d => d.Id is not null && !d.IsRestricted && d.Type.Equals("Computer", StringComparison.OrdinalIgnoreCase))
            .ToList();
        return candidates.FirstOrDefault(d => d.Name.Equals(machineName, StringComparison.OrdinalIgnoreCase))
               ?? candidates.FirstOrDefault(d => d.IsActive)
               ?? candidates.FirstOrDefault();
    }

    /// GET /me/player: 204 (empty body) means nothing is playing anywhere.
    public static SpotifyPlayback? ParsePlayback(string json)
    {
        if (string.IsNullOrWhiteSpace(json)) return null;
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        var hasItem = root.TryGetProperty("item", out var item) && item.ValueKind == JsonValueKind.Object;
        string? itemUri = hasItem ? Str(item, "uri") : null;
        string? linkedFrom = hasItem && item.TryGetProperty("linked_from", out var lf) && lf.ValueKind == JsonValueKind.Object ? Str(lf, "uri") : null;
        string? deviceId = root.TryGetProperty("device", out var device) && device.ValueKind == JsonValueKind.Object ? Str(device, "id") : null;
        var progress = root.TryGetProperty("progress_ms", out var p) && p.ValueKind == JsonValueKind.Number ? p.GetInt32() : 0;
        return new SpotifyPlayback(Bool(root, "is_playing"), progress, itemUri, Str(root, "currently_playing_type"), deviceId, linkedFrom);
    }

    /// Maps a failed player call to a PlayerException with a message fit for the island.
    public static PlayerException MapError(HttpStatusCode status, string body, TimeSpan? retryAfter)
    {
        string? message = null, reason = null;
        try
        {
            using var doc = JsonDocument.Parse(body);
            if (doc.RootElement.TryGetProperty("error", out var error) && error.ValueKind == JsonValueKind.Object)
            {
                message = Str(error, "message");
                reason = Str(error, "reason");
            }
        }
        catch (JsonException) { /* not JSON */ }

        return (int)status switch
        {
            401 => new PlayerException(PlayerErrorKind.NotAuthorized, "Spotify sign-in expired"),
            403 when reason == "PREMIUM_REQUIRED" =>
                new PlayerException(PlayerErrorKind.Unavailable, "Spotify Connect needs Spotify Premium"),
            403 => new PlayerException(PlayerErrorKind.Unavailable,
                $"Spotify refused playback control ({message ?? "forbidden"}). It needs Premium, and your account added under User Management in your Spotify developer app"),
            404 => new PlayerException(PlayerErrorKind.Unavailable,
                "Spotify isn't open on this PC. Start the Spotify app and try again"),
            429 => new PlayerException(PlayerErrorKind.Failed,
                retryAfter is { } wait
                    ? $"Spotify is rate-limiting Notchle. Try again in {Math.Max(1, (int)Math.Ceiling(wait.TotalSeconds))} s."
                    : "Spotify is rate-limiting Notchle. Try again in a minute."),
            _ => new PlayerException(PlayerErrorKind.Failed,
                $"Spotify returned HTTP {(int)status}{(message is null ? "" : $": {message}")}"),
        };
    }

    private static string WithDevice(string path, string? deviceId) =>
        deviceId is null ? path : $"{path}{(path.Contains('?') ? '&' : '?')}device_id={Uri.EscapeDataString(deviceId)}";

    private static HttpRequestMessage Put(string path, JsonNode? body) => new(HttpMethod.Put, new Uri(BaseUri, path))
    {
        // Spotify wants a Content-Length on PUT even without a body.
        Content = new StringContent(body?.ToJsonString() ?? "", Encoding.UTF8, "application/json"),
    };

    private static string? Str(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static bool Bool(JsonElement e, string name) =>
        e.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.True;
}

File: Sources/NotchleCore/Spotify/SpotifyPlayerAPI.swift
import Foundation

public struct SpotifyDevice: Sendable, Hashable {
    public var id: String?
    public var name: String
    public var type: String
    public var isActive: Bool
    public var isRestricted: Bool

    public init(id: String?, name: String, type: String, isActive: Bool = false, isRestricted: Bool = false) {
        self.id = id
        self.name = name
        self.type = type
        self.isActive = isActive
        self.isRestricted = isRestricted
    }
}

/// The parts of GET /v1/me/player the snippet timing needs.
public struct SpotifyPlayback: Sendable, Hashable {
    public var isPlaying: Bool
    public var progressMs: Int
    public var itemURI: String?
    public var currentlyPlayingType: String?
    public var deviceID: String?
    /// `item.linked_from.uri`: Spotify plays a relinked version for the account's market and
    /// reports the requested track here.
    public var linkedFromURI: String?

    public init(isPlaying: Bool, progressMs: Int, itemURI: String?, currentlyPlayingType: String? = "track",
                deviceID: String? = nil, linkedFromURI: String? = nil) {
        self.linkedFromURI = linkedFromURI
        self.isPlaying = isPlaying
        self.progressMs = progressMs
        self.itemURI = itemURI
        self.currentlyPlayingType = currentlyPlayingType
        self.deviceID = deviceID
    }

    public var position: Double { Double(progressMs) / 1000 }

    /// Whether this is `uri`, directly or as the relinked version of it.
    public func isPlayingItem(_ uri: String) -> Bool { itemURI == uri || linkedFromURI == uri }

    /// For error logs: what Spotify reported.
    public var summary: String {
        "is_playing=\(isPlaying) item=\(itemURI ?? "none") linked_from=\(linkedFromURI ?? "none") device=\(deviceID ?? "none") type=\(currentlyPlayingType ?? "none")"
    }
    public var isAd: Bool { currentlyPlayingType == "ad" }
}

/// Spotify Web API player endpoints: pure request building, response parsing and error mapping.
/// https://developer.spotify.com/documentation/web-api/reference/start-a-users-playback
public enum SpotifyPlayerAPI {
    public static let baseURL = "https://api.spotify.com/v1/"

    public static func devices() -> HTTPRequest { HTTPRequest(method: "GET", url: url("me/player/devices")) }
    /// `market=from_token` so Spotify reports `linked_from` for relinked tracks.
    public static func playbackState() -> HTTPRequest { HTTPRequest(method: "GET", url: url("me/player?market=from_token")) }
    public static func profile() -> HTTPRequest { HTTPRequest(method: "GET", url: url("me")) }

    /// PUT /me/player: move playback to this Mac's Spotify app without starting anything.
    public static func transfer(deviceID: String) -> HTTPRequest {
        put("me/player", json: ["device_ids": [deviceID], "play": false])
    }

    /// PUT /me/player/play: this one track, from positionMs, on the given device.
    public static func play(deviceID: String, trackURI: String, positionMs: Int) -> HTTPRequest {
        put("me/player/play?device_id=\(SpotifyAccounts.percentEncode(deviceID))",
            json: ["uris": [trackURI], "position_ms": max(0, positionMs)])
    }

    /// PUT /me/player/play without a body resumes the current track.
    public static func resume(deviceID: String?) -> HTTPRequest { put(withDevice("me/player/play", deviceID), json: nil) }

    public static func pause(deviceID: String?) -> HTTPRequest { put(withDevice("me/player/pause", deviceID), json: nil) }

    public static func seek(positionMs: Int, deviceID: String?) -> HTTPRequest {
        put(withDevice("me/player/seek?position_ms=\(max(0, positionMs))", deviceID), json: nil)
    }

    public static func parseDevices(_ body: Data) -> [SpotifyDevice] {
        guard let list = jsonObject(body)?["devices"] as? [[String: Any]] else { return [] }
        return list.map { d in
            SpotifyDevice(id: d["id"] as? String, name: d["name"] as? String ?? "", type: d["type"] as? String ?? "",
                          isActive: d["is_active"] as? Bool ?? false, isRestricted: d["is_restricted"] as? Bool ?? false)
        }
    }

    /// The Spotify desktop app on this Mac: a controllable "Computer" device named after this
    /// machine (the app's default device name), else the first controllable Computer.
    public static func pickLocalDevice(_ devices: [SpotifyDevice], machineName: String) -> SpotifyDevice? {
        let computers = devices.filter {
            $0.id != nil && !$0.isRestricted && $0.type.caseInsensitiveCompare("Computer") == .orderedSame
        }
        return computers.first { $0.name.caseInsensitiveCompare(machineName) == .orderedSame } ?? computers.first
    }

    /// GET /me/player: 204 (empty body) means nothing is playing anywhere → nil.
    public static func parsePlayback(_ body: Data) -> SpotifyPlayback? {
        guard let root = jsonObject(body) else { return nil }
        let item = root["item"] as? [String: Any]
        let device = root["device"] as? [String: Any]
        return SpotifyPlayback(
            isPlaying: root["is_playing"] as? Bool ?? false,
            progressMs: (root["progress_ms"] as? NSNumber)?.intValue ?? 0,
            itemURI: item?["uri"] as? String,
            currentlyPlayingType: root["currently_playing_type"] as? String,
            deviceID: device?["id"] as? String,
            linkedFromURI: (item?["linked_from"] as? [String: Any])?["uri"] as? String)
    }

    /// GET /me → display_name (falls back to the account id).
    public static func parseDisplayName(_ body: Data) -> String? {
        guard let root = jsonObject(body) else { return nil }
        if let name = root["display_name"] as? String, !name.isEmpty { return name }
        return root["id"] as? String
    }

    public static let noDeviceMessage = "Open the Spotify app on this Mac (it can stay hidden)"

    /// Maps a failed player call to a PlayerError whose text fits the notch.
    public static func mapError(status: Int, body: Data, retryAfter: String?) -> PlayerError {
        let error = jsonObject(body)?["error"] as? [String: Any]
        let message = error?["message"] as? String
        let reason = error?["reason"] as? String
        switch status {
        case 401:
            return .failed("Spotify sign-in expired. Connect Spotify again in settings")
        case 403 where reason == "PREMIUM_REQUIRED":
            return .unavailable("Spotify Connect needs Spotify Premium")
        case 403:
            return .unavailable("Spotify refused playback control (\(message ?? "forbidden")). It needs Premium, and your account added under User Management in your Spotify developer app")
        case 404:
            return .unavailable(noDeviceMessage)
        case 429:
            if let seconds = retryAfter.flatMap(Double.init) {
                return .failed("Spotify is rate-limiting Notchle. Try again in \(max(1, Int(seconds.rounded(.up)))) s.")
            }
            return .failed("Spotify is rate-limiting Notchle. Try again in a minute.")
        default:
            return .failed("Spotify returned HTTP \(status)\(message.map { ": \($0)" } ?? "")")
        }
    }

    static func jsonObject(_ body: Data) -> [String: Any]? {
        guard !body.isEmpty else { return nil }
        return (try? JSONSerialization.jsonObject(with: body)) as? [String: Any]
    }

    private static func url(_ path: String) -> URL { URL(string: baseURL + path)! }

    private static func withDevice(_ path: String, _ deviceID: String?) -> String {
        guard let deviceID else { return path }
        return "\(path)\(path.contains("?") ? "&" : "?")device_id=\(SpotifyAccounts.percentEncode(deviceID))"
    }

    private static func put(_ path: String, json: [String: Any]?) -> HTTPRequest {
        // Spotify wants a Content-Length on PUT even without a body, so always send one.
        let body = json.flatMap { try? JSONSerialization.data(withJSONObject: $0, options: [.sortedKeys]) } ?? Data()
        return HTTPRequest(method: "PUT", url: url(path), headers: ["Content-Type": "application/json"], body: body)
    }
}
